roachtest: no volatility for cast sqlsmith/setup=tpcc/setting=no-mutations failed

[(roachtest).sqlsmith/setup=tpcc/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2013639&tab=buildLog) on [master@9eea0be192499fa7d29b5eca5ed14173982b25f0](https://github.com/cockroachdb/cockroach/commits/9eea0be192499fa7d29b5eca5ed14173982b25f0):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=tpcc/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: no volatility for cast unknown::tuple{oid[], int, int, int[]}
		stmt:
		WITH
			with_68 (col_350, col_351)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									(ARRAY[3326584685:::OID,80433409:::OID,820018757:::OID], (-938321928942505224):::INT8, (-7580):::INT8, ARRAY[8950:::INT8,18770:::INT8]),
									EXISTS(
										SELECT NULL AS col_343 FROM defaultdb.public.warehouse@[0] AS tab_183 LIMIT 57:::INT8
									)
								),
								(
									(
										SELECT
											NULL AS col_344
										FROM
											defaultdb.public.warehouse@primary AS tab_184
										ORDER BY
											tab_184.w_id ASC, tab_184.w_zip
										LIMIT
											1:::INT8
									),
									not_ilike_escape(jsonb_typeof('false':::JSONB::JSONB)::STRING::STRING, e'?\x1aDKy\r\x16\f':::STRING::STRING, e'1Hc\x13$5Yj':::STRING::STRING)::BOOL
								),
								(
									(ARRAY[2929277516:::OID,1298950565:::OID,3862054642:::OID,4199588206:::OID,0:::OID,921931389:::OID], 2156249725798318035:::INT8, (-22696):::INT8, ARRAY[(-128):::INT8,127:::INT8,(-18278):::INT8,7433:::INT8,(-1):::INT8,(-13908):::INT8,(-22434):::INT8]),
									(false AND (EXISTS (WITH with_66 (col_345) AS (SELECT * FROM (VALUES (NULL), (0.4354933111764593:::FLOAT8), ((-1.4090297597753398):::FLOAT8)) AS tab_185 (col_345)), with_67 (col_346, col_347, col_348) AS (SELECT * FROM (VALUES (e'\x14\x078<=':::STRING, '\xd013479c':::BYTES, 722184144:::OID)) AS tab_186 (col_346, col_347, col_348)) SELECT cte_ref_22.col_348 AS col_349 FROM with_67 AS cte_ref_22 WHERE true) AND true))
								)
						)
							AS tab_187 (col_350, col_351)
				)
		SELECT
			cte_ref_23.col_350 AS col_352, cte_ref_23.col_351 AS col_353
		FROM
			with_68 AS cte_ref_23
		ORDER BY
			cte_ref_23.col_351 DESC,
			cte_ref_23.col_351 ASC,
			cte_ref_23.col_350 DESC,
			cte_ref_23.col_350,
			cte_ref_23.col_351 ASC,
			cte_ref_23.col_350 ASC;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpcc/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2013639&tab=artifacts#/sqlsmith/setup=tpcc/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpcc%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: no volatility for cast unknown::tuple{oid[], int, int, int[]}
```

[(roachtest).sqlsmith/setup=tpcc/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2030392&tab=buildLog) on [master@6b9c32bf8d8553be6b0540f82ce18327b88262b5](https://github.com/cockroachdb/cockroach/commits/6b9c32bf8d8553be6b0540f82ce18327b88262b5):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=tpcc/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: unexpected error from the vectorized engine: interface conversion: interface {} is *tree.DIPAddr, not *coldataext.Datum
		stmt:
		SELECT
			7338614209353257455:::INT8 AS col_235,
			tab_113.i_name AS col_236,
			array_position(ARRAY['8.79.232.88/26':::INET,'48.138.184.240/10':::INET]::INET[], ('18b5:7e2:b3b:6f35:c48:eb6a:d607:6c61/108':::INET::INET & broadcast('13.83.69.95/21':::INET::INET)::INET::INET)::INET::INET)::INT8
				AS col_237,
			tab_113.i_name AS col_238
		FROM
			defaultdb.public.item@[0] AS tab_113
		WHERE
			true;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpcc/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2030392&tab=artifacts#/sqlsmith/setup=tpcc/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpcc%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/col/coldataext/datum_vec.go
+++ b/pkg/col/coldataext/datum_vec.go
@@ -54,10 +54,11 @@ func newDatumVec(t *types.T, n int, evalCtx *tree.EvalContext) coldata.DatumVec 
 }
 
 // BinFn evaluates the provided binary function between the receiver and other.
-// dVec is the datumVec that stores either d or other (it doesn't matter which
-// one because it is used only to supply the eval context).
-func (d *Datum) BinFn(binFn *tree.BinOp, dVec, other interface{}) (tree.Datum, error) {
-	return binFn.Fn(dVec.(*datumVec).evalCtx, d.Datum, maybeUnwrapDatum(other))
+// other can either be nil, tree.Datum, or *Datum.
+func (d *Datum) BinFn(
+	binFn *tree.BinOp, evalCtx *tree.EvalContext, other interface{},
+) (tree.Datum, error) {
+	return binFn.Fn(evalCtx, d.Datum, maybeUnwrapDatum(other))
 }
 
 // CompareDatum returns the comparison between d and other. The other is

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -1489,8 +1489,7 @@ func planSelectionOperators(
 				return op, resultIdx, ct, internalMemUsedLeft, err
 			}
 			op, err := colexec.GetSelectionConstOperator(
-				lTyp, t.TypedRight().ResolvedType(), cmpOp, leftOp, leftIdx,
-				constArg, nil,
+				lTyp, t.TypedRight().ResolvedType(), cmpOp, leftOp, leftIdx, constArg, nil, /* binFn */
 			)
 			return op, resultIdx, ct, internalMemUsedLeft, err
 		}
@@ -1501,8 +1500,7 @@ func planSelectionOperators(
 			return nil, resultIdx, ct, internalMemUsed, err
 		}
 		op, err := colexec.GetSelectionOperator(
-			lTyp, ct[rightIdx], cmpOp, rightOp, leftIdx, rightIdx,
-			nil,
+			lTyp, ct[rightIdx], cmpOp, rightOp, leftIdx, rightIdx, nil, /* binFn */
 		)
 		return op, resultIdx, ct, internalMemUsedLeft + internalMemUsedRight, err
 	default:
@@ -1815,7 +1813,7 @@ func planProjectionExpr(
 		// to the input batch.
 		op, err = colexec.GetProjectionLConstOperator(
 			colmem.NewAllocator(ctx, acc, factory), left.ResolvedType(), typs[rightIdx], outputType,
-			projOp, input, rightIdx, lConstArg, resultIdx, binFn,
+			projOp, input, rightIdx, lConstArg, resultIdx, binFn, evalCtx,
 		)
 	} else {
 		var (
@@ -1863,7 +1861,7 @@ func planProjectionExpr(
 			} else {
 				op, err = colexec.GetProjectionRConstOperator(
 					colmem.NewAllocator(ctx, acc, factory), typs[leftIdx], right.ResolvedType(), outputType,
-					projOp, input, leftIdx, rConstArg, resultIdx, binFn,
+					projOp, input, leftIdx, rConstArg, resultIdx, binFn, evalCtx,
 				)
 			}
 		} else {
@@ -1882,7 +1880,7 @@ func planProjectionExpr(
 			resultIdx = len(typs)
 			op, err = colexec.GetProjectionOperator(
 				colmem.NewAllocator(ctx, acc, factory), typs[leftIdx], typs[rightIdx], outputType,
-				projOp, input, leftIdx, rightIdx, resultIdx, binFn,
+				projOp, input, leftIdx, rightIdx, resultIdx, binFn, evalCtx,
 			)
 		}
 	}

--- a/pkg/sql/colexec/execgen/cmd/execgen/overloads_base.go
+++ b/pkg/sql/colexec/execgen/cmd/execgen/overloads_base.go
@@ -923,20 +923,3 @@ func toPhysicalRepresentation(canonicalTypeFamily types.Family, width int32) str
 	// This code is unreachable, but the compiler cannot infer that.
 	return ""
 }
-
-// getDatumVecVariableName returns the variable name for a datumVec given
-// leftCol and rightCol (either of which could be "_" - meaning there is no
-// vector in scope for the corresponding side).
-// - preferRightSide determines whether we prefer the right side.
-func getDatumVecVariableName(leftCol, rightCol string, preferRightSide bool) string {
-	preferredSide, otherSide := leftCol, rightCol
-	if preferRightSide {
-		preferredSide, otherSide = rightCol, leftCol
-	}
-	switch preferredSide {
-	case "_":
-		return otherSide
-	default:
-		return preferredSide
-	}
-}

--- a/pkg/sql/colexec/execgen/cmd/execgen/overloads_bin.go
+++ b/pkg/sql/colexec/execgen/cmd/execgen/overloads_bin.go
@@ -743,27 +743,20 @@ func (c decimalIntervalCustomizer) getBinOpAssignFunc() assignFunc {
 // - leftColdataExtDatum - the variable name of the left datum element that
 // must be of *coldataext.Datum type
 // - rightDatumElem - the variable name of the right datum element which could
-// be *coldataext.Datum, tree.Datum, or nil
-// - leftCol and rightCol - same as corresponding parameters in assignFunc
-// signature
-// - datumVecOnRightSide - indicates whether we have datum-backed vector on the
-// right side (it'll be used only to supply the eval context).
-func executeBinOpOnDatums(
-	prelude, targetElem, leftColdataExtDatum, rightDatumElem, leftCol, rightCol string,
-	datumVecOnRightSide bool,
-) string {
+// be *coldataext.Datum, tree.Datum, or nil.
+func executeBinOpOnDatums(prelude, targetElem, leftColdataExtDatum, rightDatumElem string) string {
 	vecVariable, idxVariable, err := parseNonIndexableTargetElem(targetElem)
 	if err != nil {
 		return fmt.Sprintf("colexecerror.InternalError(\"%s\")", err)
 	}
 	return fmt.Sprintf(`
 			%s
-			_res, err := %s.BinFn(_overloadHelper.binFn, %s, %s)
+			_res, err := %s.BinFn(_overloadHelper.binFn, _overloadHelper.evalCtx, %s)
 			if err != nil {
 				colexecerror.ExpectedError(err)
 			}
 			%s
-		`, prelude, leftColdataExtDatum, getDatumVecVariableName(leftCol, rightCol, datumVecOnRightSide), rightDatumElem,
+		`, prelude, leftColdataExtDatum, rightDatumElem,
 		set(typeconv.DatumVecCanonicalTypeFamily, vecVariable, idxVariable, "_res"),
 	)
 }
@@ -773,7 +766,6 @@ func (c datumCustomizer) getBinOpAssignFunc() assignFunc {
 		return executeBinOpOnDatums(
 			"" /* prelude */, targetElem,
 			leftElem+".(*coldataext.Datum)", rightElem,
-			leftCol, rightCol, false, /* datumVecOnRightSide */
 		)
 	}
 }
@@ -825,7 +817,6 @@ func (c datumNonDatumCustomizer) getBinOpAssignFunc() assignFunc {
 		return executeBinOpOnDatums(
 			prelude, targetElem,
 			leftElem+".(*coldataext.Datum)", rightDatumElem,
-			leftCol, rightCol, false, /* datumVecOnRightSide */
 		)
 	}
 }
@@ -843,11 +834,7 @@ func (c nonDatumDatumCustomizer) getBinOpAssignFunc() assignFunc {
 			convertNativeToDatum(op.BinOp, c.leftCanonicalTypeFamily, leftElem, leftDatumElem),
 			leftColdataExtDatum, leftDatumElem,
 		)
-		return executeBinOpOnDatums(
-			prelude, targetElem,
-			leftColdataExtDatum, rightElem,
-			leftCol, rightCol, true, /* datumVecOnRightSide */
-		)
+		return executeBinOpOnDatums(prelude, targetElem, leftColdataExtDatum, rightElem)
 	}
 }
 

--- a/pkg/sql/colexec/execgen/cmd/execgen/overloads_bin.go
+++ b/pkg/sql/colexec/execgen/cmd/execgen/overloads_bin.go
@@ -127,13 +127,6 @@ func registerBinOpOutputTypes() {
 			binOpOutputTypes[binOp][typePair{types.DecimalFamily, anyWidth, types.IntFamily, intWidth}] = types.Decimal
 			binOpOutputTypes[binOp][typePair{types.IntFamily, intWidth, types.DecimalFamily, anyWidth}] = types.Decimal
 		}
-
-		for _, compatibleFamily := range compatibleCanonicalTypeFamilies[typeconv.DatumVecCanonicalTypeFamily] {
-			for _, width := range supportedWidthsByCanonicalTypeFamily[compatibleFamily] {
-				binOpOutputTypes[binOp][typePair{typeconv.DatumVecCanonicalTypeFamily, anyWidth, compatibleFamily, width}] = types.Any
-				binOpOutputTypes[binOp][typePair{compatibleFamily, width, typeconv.DatumVecCanonicalTypeFamily, anyWidth}] = types.Any
-			}
-		}
 	}
 	// There is a special case for division with integers; it should have a
 	// decimal result.
@@ -156,6 +149,26 @@ func registerBinOpOutputTypes() {
 	binOpOutputTypes[tree.Plus][typePair{types.TimestampTZFamily, anyWidth, types.IntervalFamily, anyWidth}] = types.TimestampTZ
 	binOpOutputTypes[tree.Minus][typePair{types.TimestampTZFamily, anyWidth, types.IntervalFamily, anyWidth}] = types.TimestampTZ
 	binOpOutputTypes[tree.Plus][typePair{types.IntervalFamily, anyWidth, types.TimestampTZFamily, anyWidth}] = types.TimestampTZ
+	for _, compatibleFamily := range []types.Family{
+		types.IntFamily,      // types.Date + types.Time
+		types.IntervalFamily, // types.Time + types.Interval
+	} {
+		for _, width := range supportedWidthsByCanonicalTypeFamily[compatibleFamily] {
+			binOpOutputTypes[tree.Plus][typePair{typeconv.DatumVecCanonicalTypeFamily, anyWidth, compatibleFamily, width}] = types.Any
+			binOpOutputTypes[tree.Plus][typePair{compatibleFamily, width, typeconv.DatumVecCanonicalTypeFamily, anyWidth}] = types.Any
+		}
+	}
+	for _, compatibleFamily := range []types.Family{
+		typeconv.DatumVecCanonicalTypeFamily, // types.INet - types.INet
+		types.IntFamily,                      // types.Date - types.Time
+		types.IntervalFamily,                 // types.Time - types.Interval
+		types.BytesFamily,                    // types.Jsonb - types.String
+	} {
+		for _, width := range supportedWidthsByCanonicalTypeFamily[compatibleFamily] {
+			binOpOutputTypes[tree.Minus][typePair{typeconv.DatumVecCanonicalTypeFamily, anyWidth, compatibleFamily, width}] = types.Any
+			binOpOutputTypes[tree.Minus][typePair{compatibleFamily, width, typeconv.DatumVecCanonicalTypeFamily, anyWidth}] = types.Any
+		}
+	}
 
 	// Other arithmetic binary operators.
 	for _, binOp := range []tree.BinaryOperator{tree.FloorDiv, tree.Mod, tree.Pow} {

--- a/pkg/sql/colexec/execgen/cmd/execgen/overloads_cmp.go
+++ b/pkg/sql/colexec/execgen/cmd/execgen/overloads_cmp.go
@@ -331,9 +331,19 @@ func (c intervalCustomizer) getCmpOpCompareFunc() compareFunc {
 	}
 }
 
+// getDatumVecVariableName returns the variable name for a datumVec given
+// leftCol and rightCol (either of which could be "_" - meaning there is no
+// vector in scope for the corresponding side).
+func getDatumVecVariableName(leftCol, rightCol string) string {
+	if leftCol == "_" {
+		return rightCol
+	}
+	return leftCol
+}
+
 func (c datumCustomizer) getCmpOpCompareFunc() compareFunc {
 	return func(targetElem, leftElem, rightElem, leftCol, rightCol string) string {
-		datumVecVariableName := getDatumVecVariableName(leftCol, rightCol, false /* preferRightSide */)
+		datumVecVariableName := getDatumVecVariableName(leftCol, rightCol)
 		return fmt.Sprintf(`
 			%s = %s.(*coldataext.Datum).CompareDatum(%s, %s)
 		`, targetElem, leftElem, datumVecVariableName, rightElem)

--- a/pkg/sql/colexec/proj_const_ops_tmpl.go
+++ b/pkg/sql/colexec/proj_const_ops_tmpl.go
@@ -23,6 +23,7 @@ import (
 	"context"
 
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
+	"github.com/cockroachdb/cockroach/pkg/col/coldataext"
 	"github.com/cockroachdb/cockroach/pkg/col/typeconv"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexec/execgen"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecbase"
@@ -217,6 +218,7 @@ func GetProjection_CONST_SIDEConstOperator(
 	constArg tree.Datum,
 	outputIdx int,
 	binFn *tree.BinOp,
+	evalCtx *tree.EvalContext,
 ) (colexecbase.Operator, error) {
 	input = newVectorTypeEnforcer(allocator, input, outputType, outputIdx)
 	projConstOpBase := projConstOpBase{
@@ -224,7 +226,7 @@ func GetProjection_CONST_SIDEConstOperator(
 		allocator:      allocator,
 		colIdx:         colIdx,
 		outputIdx:      outputIdx,
-		overloadHelper: overloadHelper{binFn: binFn},
+		overloadHelper: overloadHelper{binFn: binFn, evalCtx: evalCtx},
 	}
 	var (
 		c   interface{}
@@ -245,6 +247,7 @@ func GetProjection_CONST_SIDEConstOperator(
 		case tree._NAME:
 			switch typeconv.TypeFamilyToCanonicalTypeFamily(leftType.Family()) {
 			// {{range .LeftFamilies}}
+			// {{$leftFamilyStr := .LeftCanonicalFamilyStr}}
 			case _LEFT_CANONICAL_TYPE_FAMILY:
 				switch leftType.Width() {
 				// {{range .LeftWidths}}
@@ -258,7 +261,20 @@ func GetProjection_CONST_SIDEConstOperator(
 							return &_OP_CONST_NAME{
 								projConstOpBase: projConstOpBase,
 								// {{if _IS_CONST_LEFT}}
+								// {{if eq $leftFamilyStr "typeconv.DatumVecCanonicalTypeFamily"}}
+								// {{/*
+								//     Binary operations are evaluated using coldataext.Datum.BinFn
+								//     method which requires that we have *coldataext.Datum on the
+								//     left, so we create that at the operator construction time to
+								//     avoid runtime conversion. Note that when the constant is on
+								//     the right side, then the left element necessarily comes from
+								//     the vector and will be of the desired type, so no additional
+								//     work is needed.
+								// */}}
+								constArg: &coldataext.Datum{Datum: c.(tree.Datum)},
+								// {{else}}
 								constArg: c.(_L_GO_TYPE),
+								// {{end}}
 								// {{else}}
 								constArg: c.(_R_GO_TYPE),
 								// {{end}}
@@ -279,6 +295,7 @@ func GetProjection_CONST_SIDEConstOperator(
 		case tree._NAME:
 			switch typeconv.TypeFamilyToCanonicalTypeFamily(leftType.Family()) {
 			// {{range .LeftFamilies}}
+			// {{$leftFamilyStr := .LeftCanonicalFamilyStr}}
 			case _LEFT_CANONICAL_TYPE_FAMILY:
 				switch leftType.Width() {
 				// {{range .LeftWidths}}
@@ -292,7 +309,27 @@ func GetProjection_CONST_SIDEConstOperator(
 							return &_OP_CONST_NAME{
 								projConstOpBase: projConstOpBase,
 								// {{if _IS_CONST_LEFT}}
+								// {{if eq $leftFamilyStr "typeconv.DatumVecCanonicalTypeFamily"}}
+								// {{/*
+								//     Comparison operations are evaluated using
+								//     coldataext.Datum.CompareDatum method which requires that we
+								//     have *coldataext.Datum on the left, so we create that at the
+								//     operator construction time to avoid runtime conversion. Note
+								//     that when the constant is on the right side, then the left
+								//     element necessarily comes from the vector and will be of the
+								//     desired type, so no additional work is needed.
+								//     Note: although it appears as if the comparison expressions
+								//     supported by this template (EQ, NE, LT, GT, LE, GE) are
+								//     normalized so that the constant is on the right side, we
+								//     choose to be safe and perform the conversion here.
+								//     TODO(yuzefovich): this appears to be a dead code, look into
+								//     whether we can skip generation of the code for the case when
+								//     the constant is on the left.
+								// */}}
+								constArg: &coldataext.Datum{Datum: c.(tree.Datum)},
+								// {{else}}
 								constArg: c.(_L_GO_TYPE),
+								// {{end}}
 								// {{else}}
 								constArg: c.(_R_GO_TYPE),
 								// {{end}}

--- a/pkg/sql/colexec/proj_non_const_ops_tmpl.go
+++ b/pkg/sql/colexec/proj_non_const_ops_tmpl.go
@@ -247,6 +247,7 @@ func GetProjectionOperator(
 	col2Idx int,
 	outputIdx int,
 	binFn *tree.BinOp,
+	evalCtx *tree.EvalContext,
 ) (colexecbase.Operator, error) {
 	input = newVectorTypeEnforcer(allocator, input, outputType, outputIdx)
 	projOpBase := projOpBase{
@@ -255,7 +256,7 @@ func GetProjectionOperator(
 		col1Idx:        col1Idx,
 		col2Idx:        col2Idx,
 		outputIdx:      outputIdx,
-		overloadHelper: overloadHelper{binFn: binFn},
+		overloadHelper: overloadHelper{binFn: binFn, evalCtx: evalCtx},
 	}
 
 	switch op.(type) {

--- a/pkg/sql/colexec/projection_ops_test.go
+++ b/pkg/sql/colexec/projection_ops_test.go
@@ -165,8 +165,8 @@ func TestGetProjectionConstOperator(t *testing.T) {
 	constArg := tree.NewDFloat(tree.DFloat(constVal))
 	outputIdx := 5
 	op, err := GetProjectionRConstOperator(
-		testAllocator, types.Float, types.Float, types.Float,
-		binOp, input, colIdx, constArg, outputIdx, nil,
+		testAllocator, types.Float, types.Float, types.Float, binOp, input,
+		colIdx, constArg, outputIdx, nil /* binFn */, nil, /* evalCtx */
 	)
 	if err != nil {
 		t.Error(err)
@@ -194,8 +194,8 @@ func TestGetProjectionConstMixedTypeOperator(t *testing.T) {
 	constArg := tree.NewDInt(tree.DInt(constVal))
 	outputIdx := 5
 	op, err := GetProjectionRConstOperator(
-		testAllocator, types.Int, types.Int2, types.Int,
-		binOp, input, colIdx, constArg, outputIdx, nil,
+		testAllocator, types.Int, types.Int2, types.Int, binOp, input, colIdx,
+		constArg, outputIdx, nil /* binFn */, nil, /* evalCtx */
 	)
 	if err != nil {
 		t.Error(err)
@@ -319,8 +319,8 @@ func TestGetProjectionOperator(t *testing.T) {
 	col2Idx := 7
 	outputIdx := 9
 	op, err := GetProjectionOperator(
-		testAllocator, typ, typ, types.Int2,
-		binOp, input, col1Idx, col2Idx, outputIdx, nil,
+		testAllocator, typ, typ, types.Int2, binOp, input, col1Idx, col2Idx,
+		outputIdx, nil /* binFn */, nil, /* evalCtx */
 	)
 	if err != nil {
 		t.Error(err)

--- a/pkg/sql/colexec/utils.go
+++ b/pkg/sql/colexec/utils.go
@@ -47,6 +47,7 @@ var (
 type overloadHelper struct {
 	tmpDec1, tmpDec2 apd.Decimal
 	binFn            *tree.BinOp
+	evalCtx          *tree.EvalContext
 }
 
 // makeWindowIntoBatch updates windowedBatch so that it provides a "window"

--- a/pkg/sql/logictest/testdata/logic_test/vectorize_overloads
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize_overloads
@@ -561,3 +561,43 @@ EXPLAIN (VEC) SELECT _time + _date FROM many_types
 │
 └ Node 1
   └ *rowexec.tableReader
+
+# Regression #50261 (not handling constant datum-backed values on the left
+# correctly).
+statement error pq: cannot AND inet values of different sizes
+SELECT '18b5:7e2:b3b:6f35:c48:eb6a:d607:6c61/108':::INET::INET & broadcast('13.83.69.95/21':::INET::INET)::INET::INET FROM many_types WHERE _bool
+
+query T
+EXPLAIN (VEC) SELECT '[2, "hi", {"b": ["bar", {"c": 4}]}]'::jsonb -> _int FROM many_types
+----
+│
+└ Node 1
+  └ *colexec.projJSONFetchValDatumConstInt64Op
+    └ *colfetcher.colBatchScan
+
+query T rowsort
+SELECT '[2, "hi", {"b": ["bar", {"c": 4}]}]'::jsonb -> _int FROM many_types
+----
+NULL
+NULL
+NULL
+{"b": ["bar", {"c": 4}]}
+
+# Check that the comparison expressions with the constant on the left are
+# handled as well (such expressions are normalized so that the constant ends up
+# on the right side).
+query T
+EXPLAIN (VEC) SELECT B'11' <> _varbit FROM many_types
+----
+│
+└ Node 1
+  └ *colexec.projNEDatumDatumConstOp
+    └ *colfetcher.colBatchScan
+
+query B rowsort
+SELECT B'11' >= _varbit FROM many_types
+----
+NULL
+true
+true
+false

